src/scheduled_tasks.rs

```diff
@@ -100,64 +93,93 @@ fn startup_jobs(db_url: &str) {
+    naive_now() - chrono::Duration::days(7)
   } else {
     info!("Updating hot ranks for all history...");
-  }
+    NaiveDateTime::from_timestamp_opt(0, 0).expect("0 timestamp creation")
+  };
 
-  match post_update
-    .set((
-      post_aggregates::hot_rank.eq(hot_rank(post_aggregates::score, post_aggregates::published)),
-      post_aggregates::hot_rank_active.eq(hot_rank(
-        post_aggregates::score,
-        post_aggregates::newest_comment_time_necro,
-      )),
-    ))
-    .execute(conn)
-  {
-    Ok(_) => {}
-    Err(e) => {
-      error!("Failed to update post_aggregates hot_ranks: {}", e)
-    }
-  }
+  process_hot_ranks_in_batches(
+    conn,
+    "post_aggregates",
+    "SET hot_rank = hot_rank(a.score, a.published),
+         hot_rank_active = hot_rank(a.score, a.newest_comment_time_necro)",
+    process_start_time,
+  );
+
+  process_hot_ranks_in_batches(
+    conn,
+    "comment_aggregates",
+    "SET hot_rank = hot_rank(a.score, a.published)",
+    process_start_time,
+  );
+
+  process_hot_ranks_in_batches(
+    conn,
+    "community_aggregates",
+    "SET hot_rank = hot_rank(a.subscribers, a.published)",
+    process_start_time,
+  );
+
+  info!("Finished hot ranks update!");
+}
 
-  match comment_update
-    .set(comment_aggregates::hot_rank.eq(hot_rank(
-      comment_aggregates::score,
-      comment_aggregates::published,
-    )))
-    .execute(conn)
-  {
-    Ok(_) => {}
-    Err(e) => {
-      error!("Failed to update comment_aggregates hot_ranks: {}", e)
-    }
-  }
+#[derive(QueryableByName)]
+struct HotRanksUpdateResult {
+  #[diesel(sql_type = Timestamp)]
+  published: NaiveDateTime,
+}
 
-  match community_update
-    .set(community_aggregates::hot_rank.eq(hot_rank(
-      community_aggregates::subscribers,
-      community_aggregates::published,
-    )))
-    .execute(conn)
-  {
-    Ok(_) => {
-      info!("Done.");
-    }
-    Err(e) => {
-      error!("Failed to update community_aggregates hot_ranks: {}", e)
+/// Runs the hot rank update query in batches until all ids have been processed.
+/// In `set_clause`, "a" will refer to the current aggregates table.
+/// Locked rows are skipped in order to prevent deadlocks (they will likely get updated on the next
+/// run)
+fn process_hot_ranks_in_batches(
+  conn: &mut PgConnection,
+  table_name: &str,
+  set_clause: &str,
+  process_start_time: NaiveDateTime,
+) {
+  let update_batch_size = 1000; // Bigger batches than this tend to cause seq scans
+  let mut previous_batch_last_published = process_start_time;
+  loop {
+    // Raw `sql_query` is used as a performance optimization - Diesel does not support doing this
+    // in a single query (neither as a CTE, nor using a subquery)
+    let result = sql_query(format!(
+      r#"WITH batch AS (SELECT a.id
+               FROM {aggregates_table} a
+               WHERE a.published > $1
+               ORDER BY a.published
```

Comment:
The indexes you added were `published desc`, would be good to use that. Or if that doesn't work because of the incrementing `previous_batch_last_published` used here, then make those indexes `asc`

Reply:
I went with `desc` indexes for the two new ones, because the existing one on `post_aggregates` already had a `desc` index, so I thought it looked more uniform 😃

AFAIK there should be no performance difference between forward and backward index scans, do you know differently?

Btw, this query is definitely using the index:

```
Update on post_aggregates a  (cost=83.81..2078.90 rows=1000 width=38) (actual time=0.300..172.142 rows=1000 loops=1)
  CTE batch
    ->  Limit  (cost=0.42..83.40 rows=1000 width=88) (actual time=0.062..8.010 rows=1000 loops=1)
          ->  LockRows  (cost=0.42..2630.48 rows=31697 width=88) (actual time=0.060..7.637 rows=1000 loops=1)
                ->  Index Scan Backward using idx_post_aggregates_published on post_aggregates a_1  (cost=0.42..2313.51 rows=31697 width=88) (actual time=0.044..5.588 rows=1000 loops=1)
                      Index Cond: (published > (now() - '7 days'::interval))
  ->  Nested Loop  (cost=0.42..1995.50 rows=1000 width=38) (actual time=0.256..138.719 rows=1000 loops=1)
        ->  CTE Scan on batch  (cost=0.00..20.00 rows=1000 width=32) (actual time=0.077..11.958 rows=1000 loops=1)
        ->  Index Scan using post_aggregates_pkey on post_aggregates a  (cost=0.42..1.72 rows=1 width=26) (actual time=0.007..0.007 rows=1 loops=1000)
              Index Cond: (id = batch.id)
Planning Time: 0.463 ms
Execution Time: 172.697 ms
```

Reply:
DESC and ASC indexes can both be used for either direction of query, they have a slight performance difference that's afaik only really relevant when the index is not in RAM and you're scanning through a large amount of rows in the index (it determines the ordering on disk where fetching forward is more efficient than fetching backwards due to prefetch)